### PR TITLE
allow list of search patterns in testFiles config option

### LINKED ISSUE #5401: Allow passing a list of glob patterns in testFiles config option
Cypress v3.4.1

There is a configuration option `testFiles` in https://on.cypress.io/configuration that controls which spec files are found. 
Let's say I have the following files

```
cypress/
	integration/
		foo.js
		bar.js
		baz.js
```

If I want to only have "foo.js" file I can do the following

```json
{
  "testFiles": "foo.js"
}
```

But how do I pick "foo.js" AND "bar.js"? 

I would be very nice to allow `testFiles` to be a single string or a list of strings and match each filename against a list.

### PR DESCRIPTION
**Pull request by bahmutov**
- Closes #5401 

### User facing changelog

- [x] allows a list of specs when searching for test files via config argument `testFiles` which follows the principle in `ignoreTestFiles` config option

### Additional details

Using [`glob`](https://github.com/isaacs/node-glob) to select several groups of spec names is not obvious. For example to select both `.md` and `*spec.js` files one would need to write without spaces

```json
{
  "testFiles": "{*.md,*spec.js}"
}
```

With this change, user could write

```json
{
  "testFiles": ["*.md", "*spec.js"]
}
```

which to me looks more natural and easier to remember. Also added debug log messages during `DEBUG=cypress:server:specs` that show how the search happens in case on needs to debug it.

### PR Tasks

- [x] Have tests been added/updated?
- [X] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? https://github.com/cypress-io/cypress-documentation/pull/2184
- [x] Have API changes been updated in the [`type definitions`](cli/types/index.d.ts)?
- [x] Have new configuration options been added to the [`cypress.schema.json`](cli/schema/cypress.schema.json)?

### CODE REVIEW

File: packages/server/test/unit/specs_spec.coffee [hunk 1]
```diff
@@ -56,23 +57,48 @@ describe "lib/util/specs", ->
...
+
     it "filters using specPattern", ->
       config.get(FixturesHelper.projectPath("various-file-types"))
       .then (cfg) ->
-        specsUtil.find(cfg, [
+        specPattern = [
           path.join(cfg.projectRoot, "cypress", "integration", "js_spec.js")
-        ])
+        ]
+        specsUtil.find(cfg, specPattern)
```
Comment (bahmutov) [inline packages/server/test/unit/specs_spec.coffee]: being a little bit clearer on the arguments

File: packages/server/test/unit/specs_spec.coffee [hunk 2]
```diff
@@ -56,23 +57,48 @@ describe "lib/util/specs", ->
...
       .then (files) ->
         expect(files.length).to.equal(1)
         expect(files[0].name).to.equal("js_spec.js")
 
     it "filters using specPattern as array of glob patterns", ->
       config.get(FixturesHelper.projectPath("various-file-types"))
       .then (cfg) ->
-        specsUtil.find(cfg, [
+        debug("test config testFiles is %o", cfg.testFiles)
```
Comment (bahmutov) [inline packages/server/test/unit/specs_spec.coffee]: adding a few debug statements to the test itself is useful when iterating on the test and understanding what is going on

File: packages/server/lib/util/specs.js [hunk 3]
```diff
@@ -132,17 +142,31 @@ const find = function (config, specPattern) {
...
 
-  //# filter out anything that matches our
-  //# ignored test files glob
-  .filter(doesNotMatchAllIgnoredPatterns)
-  .filter(matchesSpecPattern)
-  .map(setNameParts)
-  .tap((files) => {
-    return debug('found %d spec files: %o', files.length, files)
-  })
+  return Promise.mapSeries(testFilesPatterns, findOnePattern).then(_.flatten)
```
Comment (bahmutov) [inline packages/server/lib/util/specs.js]: find matching spec files one pattern at a time and then merging them

File: cli/schema/cypress.schema.json [hunk 4]
```diff
@@ -45,7 +45,7 @@
     "testFiles": {
       "type": "string",
       "default": "**/*.*",
-      "description": "A String glob pattern of the test files to load"
+      "description": "A String glob pattern of the test files to load. See https://on.cypress.io/configuration#Global"
```
Comment (bahmutov) [inline cli/schema/cypress.schema.json]: TODO: needs to be updated before merging

File: packages/server/test/integration/cypress_spec.coffee [hunk 5]
```diff
@@ -336,6 +336,28 @@ describe "lib/cypress", ->
...
+        @expectExitWith(0)
+
+    it "runs project by limiting spec files via config.testFiles as an array of string glob patterns", ->
+      cypress.start(["--run-project=#{@todosPath}", "--config='[\"testFiles=**/test2.coffee\",\"**/test1.js\"]'"])
+      .then =>
+        expect(browsers.open).to.be.calledWithMatch(ELECTRON_BROWSER, {url: "http://localhost:8888/__/#/tests/integration/test2.coffee"})
+      .then =>
+        expect(browsers.open).to.be.calledWithMatch(ELECTRON_BROWSER, {url: "http://localhost:8888/__/#/tests/integration/test1.js"})
+        @expectExitWith(0)
+
```
Comment (andrew-codes) [inline packages/server/test/integration/cypress_spec.coffee]: @bahmutov per the commit message, I'll echo it here, I think these fail for similar reasons as #4896. There is an outstanding PR for this issue. I can review the PR, get it merged when ready, and then update this branch to see if it resolves this issue.

File: cli/types/cypress-npm-api.d.ts [hunk 6]
```diff
@@ -43,7 +43,7 @@ declare module 'cypress' {
      */
     reporter: string,
     /**
-     * A String glob pattern of the test files to load.
+     * A String or Array of string glob pattern of the test files to load.
```
Comment (bahmutov) [inline cli/types/cypress-npm-api.d.ts]: got to update the type here as well
```
testFiles: string | string[]
```

File: packages/server/test/integration/cypress_spec.coffee (hunk 5 above)
Comment (bahmutov) [reply packages/server/test/integration/cypress_spec.coffee]: oh my

Comment (andrew-codes) [reply packages/server/test/integration/cypress_spec.coffee]: I retract my assessment. It looks like we do **not** need the functionality in that PR. Furthermore, supporting the syntax `--config=testFiles=["**/*.js", "**/*.coffee"]` is how other mutli-value config values work; such as ignoreTestFiles. I do not see any documentation or elsewhere in tests indicating that the format, `--config=testFiles=**/*.js,**/*.coffee`, is actually supported.

File: packages/server/test/integration/cypress_spec.coffee [hunk 7]
```diff
@@ -336,6 +336,20 @@ describe "lib/cypress", ->
...
+    it "runs project by limiting spec files via config.testFiles string glob pattern", ->
+      cypress.start(["--run-project=#{@todosPath}", "--config=testFiles=#{@todosPath}/tests/test2.coffee"])
+      .then =>
+        expect(browsers.open).to.be.calledWithMatch(ELECTRON_BROWSER, {url: "http://localhost:8888/__/#/tests/integration/test2.coffee"})
+        @expectExitWith(0)
+
+    it "runs project by limiting spec files via config.testFiles as a JSON array of string glob patterns", ->
+      cypress.start(["--run-project=#{@todosPath}", "--config=testFiles=[\"**/test2.coffee\",\"**/test1.js\"]"])
+      .then =>
+        expect(browsers.open).to.be.calledWithMatch(ELECTRON_BROWSER, {url: "http://localhost:8888/__/#/tests/integration/test2.coffee"})
```
Comment (bahmutov) [inline packages/server/test/integration/cypress_spec.coffee]: nice

File: packages/server/test/unit/config_spec.coffee [hunk 8]
```diff
@@ -322,9 +322,18 @@ describe "lib/config", ->
           @setup({testFiles: "**/*.coffee"})
           @expectValidationPasses()
 
-        it "fails if not a string", ->
+        it "passes if an array of strings", ->
+          @setup({ignoreTestFiles: ["**/*.coffee", "**/*.jsx"]})
```
Comment (bahmutov) [inline packages/server/test/unit/config_spec.coffee]: you mean `testFiles` here right ;)

File: packages/server/test/unit/config_spec.coffee [hunk 9]
```diff
@@ -322,9 +322,18 @@ describe "lib/config", ->
...
+          @expectValidationPasses()
+
+        it "fails if not a string or array", ->
           @setup({testFiles: 42})
-          @expectValidationFails("be a string")
+          @expectValidationFails("be a string or an array of string")
+
+        it "fails if not an array of strings", ->
+          @setup({ignoreTestFiles: [5]})
```
Comment (bahmutov) [inline packages/server/test/unit/config_spec.coffee]: you mean `testFiles`

**Review by bahmutov — COMMENTED** [review]
see two comments in tests

File: packages/server/test/unit/config_spec.coffee (hunk 8 above)
Comment (andrew-codes) [reply packages/server/test/unit/config_spec.coffee]: Yup, let me change these two places within `config_spec`

File: cli/schema/cypress.schema.json [hunk 10]
```diff
@@ -45,7 +45,7 @@
     "testFiles": {
       "type": "string",
       "default": "**/*.*",
-      "description": "A String glob pattern of the test files to load"
+      "description": "A String or Array of string glob pattern of the test files to load. See https://on.cypress.io/configuration#Global"
```
Comment (Bkucera) [inline cli/schema/cypress.schema.json]: doesn't sound right as is, maybe something like
```suggestion
      "description": "A String or Array of glob patterns of test files to load. See https://on.cypress.io/configuration#Global"
```

File: cli/schema/cypress.schema.json [hunk 11]
```diff
@@ -45,7 +45,7 @@
     "testFiles": {
       "type": "string",
```
Comment (Bkucera) [inline cli/schema/cypress.schema.json]:
```suggestion
      "type": [
        "string",
        "array"
      ],
```

File: packages/server/test/integration/cypress_spec.coffee [hunk 12]
```diff
@@ -758,6 +772,7 @@ describe "lib/cypress", ->
 
           @expectExitWith(0)
 
+
```
Comment (Bkucera) [inline packages/server/test/integration/cypress_spec.coffee]:
```suggestion
```

**Review by kuceb — CHANGES_REQUESTED** [review]

File: cli/schema/cypress.schema.json (hunk 10 above)
Comment (andrew-codes) [reply cli/schema/cypress.schema.json]: Agreed and changed.

File: cli/schema/cypress.schema.json (hunk 11 above)
Comment (andrew-codes) [reply cli/schema/cypress.schema.json]: Good catch! I have updated this.

File: packages/server/test/integration/cypress_spec.coffee [hunk 13]
```diff
@@ -336,6 +336,28 @@ describe "lib/cypress", ->
...
+    it "runs project by limiting spec files via config.testFiles as an array of string glob patterns", ->
+      cypress.start(["--run-project=#{@todosPath}", "--config=testFiles=**/test2.coffee,**/test1.js"])
+      .then =>
+        expect(browsers.open).to.be.calledWithMatch(ELECTRON_BROWSER, {url: "http://localhost:8888/__/#/tests/integration/test2.coffee"})
+      .then =>
+        expect(browsers.open).to.be.calledWithMatch(ELECTRON_BROWSER, {url: "http://localhost:8888/__/#/tests/integration/test1.js"})
+        @expectExitWith(0)
+
+    it "runs project by limiting spec files via config.testFiles as an array of string glob patterns", ->
+      cypress.start(["--run-project=#{@todosPath}", "--config='[\"testFiles=**/test2.coffee\",\"**/test1.js\"]'"])
```
Comment (Bkucera) [inline packages/server/test/integration/cypress_spec.coffee]: what is this syntax `--config='["testFiles=`? there is a `[` after `--config=`, which I do not see anywhere else in the tests

**Review by kuceb — CHANGES_REQUESTED** [review]

File: packages/server/test/integration/cypress_spec.coffee (hunk 13 above)
Comment (andrew-codes) [reply packages/server/test/integration/cypress_spec.coffee]: I was not sure what the syntax was supposed to be, but found a similar multi-value configuration option, `blacklistHosts` [here](https://github.com/cypress-io/cypress/blob/92b91fe514e5ff6286b4d4e26d2df23062bdf869/packages/server/test/unit/args_spec.coffee#L140-L148), so I followed that pattern. In the link, it appears that the multi-value value of blacklistHosts is simply JSON.stringified as the CLI option.

If this is not correct, we can change it. I may have found or looked in the wrong places to find the syntax.

The output of this discussion will also impact https://github.com/cypress-io/cypress-documentation/pull/2184

Comment (andrew-codes) [reply packages/server/test/integration/cypress_spec.coffee]: I think the consensus is to follow the same syntax as `blacklistHosts`; in which case this is the correct syntax: `--config=testFiles=["glob1","glob2"]`.

Note that `spec` and `tag` are run with a slightly different syntax, so we may want to reconsider the syntax across all multi-value values in the future.

Comment (Bkucera) [reply packages/server/test/integration/cypress_spec.coffee]: notice in that example you've linked, the `[` does not included in the args in the cli

Comment (andrew-codes) [reply packages/server/test/integration/cypress_spec.coffee]: In hindsight, I don't think my example and how it was conveyed was very clear. Unless I am misunderstanding you, I think the CLI args are: something like this:

`--config=pageLoadTimeout=10000,waitForAnimations=false,hosts={"foobar.com":"127.0.0.1"},blacklistHosts=["one.com","www.two.io"]`

My reasoning is outlined, hopefully more clearly, below:

- `blacklistHosts` is an array [here](https://github.com/cypress-io/cypress/blob/92b91fe514e5ff6286b4d4e26d2df23062bdf869/packages/server/test/unit/args_spec.coffee#L128)
- it is turned into JSON [here](https://github.com/cypress-io/cypress/blob/92b91fe514e5ff6286b4d4e26d2df23062bdf869/packages/server/test/unit/args_spec.coffee#L140)
- that JSON array string is used for the blacklistHosts CLI command [here](https://github.com/cypress-io/cypress/blob/92b91fe514e5ff6286b4d4e26d2df23062bdf869/packages/server/test/unit/args_spec.coffee#L148)
- the `join()` is joining all the CLI arguments; one of which is blacklistHosts which is `blacklistHosts=["one.com", "www.two.io"]`

File: packages/server/test/integration/cypress_spec.coffee (hunk 7 above)
Comment (andrew-codes) [reply packages/server/test/integration/cypress_spec.coffee]: ty

File: packages/server/test/integration/cypress_spec.coffee (hunk 13 above)
Comment (andrew-codes) [reply packages/server/test/integration/cypress_spec.coffee]: Per our discussion, I understand what you are saying now. Resolving this comment and updating code accordingly.

File: cli/schema/cypress.schema.json [hunk 14]
```diff
@@ -43,9 +43,12 @@
...
     },
     "testFiles": {
-      "type": "string",
+      "type": [
+        "string",
+        "array"
+      ],
       "default": "**/*.*",
-      "description": "A String glob pattern of the test files to load"
+      "description": "A String or Array of glob patterns of the test files to load. See https://on.cypress.io/configuration#Global"
```
Comment (Bkucera) [inline cli/schema/cypress.schema.json]:
```suggestion
      "description": "A String or Array of string glob patterns of the test files to load. See https://on.cypress.io/configuration#Global"
```

**Review by kuceb — DISMISSED** [review]
accept that wording change if you want. but LGTM

**Comment by Bkucera** [comment]
make sure to merge in develop and rebuild after we get this circle zip file size figured out

**Review by kuceb — APPROVED** [review]

**Pull request merged**